### PR TITLE
chore: add deadlock detector to actor system

### PR DESCRIPTION
**Pull request by stoksc**
## Description
This change adds a deadlock detector that can be enabled at runtime to check for a cycle of blocking asks.

## Test Plan
- [x] write test cases

## Commentary (optional)
Test cases that test deadlocks are really hard to write, so this does some _sort of_ gross sleeping but it's valid.

[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234

### CODE REVIEW

**Comment by stoksc** [comment]
This PR is blocked on #1327.

**Review by shiyuann — COMMENTED** [review]
is the plan to add deadlock detected to just e2e tests or also production builds?

File: master/pkg/actor/address.go [hunk 1]
```diff
@@ -103,3 +104,10 @@ func (a *Address) UnmarshalText(text []byte) error {
 	a.path = string(text)
 	return nil
 }
+
+// ID implements the graph.Node interface.
```
[the remark this answers is not in the document]
Comment (stoksc) [reply master/pkg/actor/address.go:108]: Yeah, it's for using the gonum package's graph impl.

File: master/pkg/actor/trace.go [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+			LogSpans: true,
+		},
+	}.NewTracer()
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "failed to init Jaeger")
+	}
+	return tracer, closer, nil
+}
+
+// Actors that are too noisy to be worth tracing.
```
[the remark this answers is not in the document]
Comment (stoksc) [reply master/pkg/actor/trace.go:108]: this is part of the tracing pr that already landed, but it just is to not trace notify actors that generate too many traces to be useful.

**Comment by stoksc** [comment]
@shiyuann this would just be for dev builds, to check if you were running into issues. not really pushing to merge very hard; it's been open a month now and i haven't needed to use it. in remove steps it would've been useful to me, but only once.

**Comment by stoksc** [comment]
Going to close this. If in the future we find we need it, I can reopen an de-conflict then.

**Pull request closed without merging**